src/js/dgram.js

```diff
@@ -0,0 +1,468 @@
+/* Copyright 2016 Samsung Electronics Co., Ltd.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+var EventEmitter = require('events').EventEmitter;
+var util = require('util');
+
+var UDP = process.binding(process.binding.udp);
+
+var BIND_STATE_UNBOUND = 0;
+var BIND_STATE_BINDING = 1;
+var BIND_STATE_BOUND = 2;
+
+// lazily loaded
+var dns = null;
+
+function lookup(address, family, callback) {
+  if (!dns)
+    dns = require('dns');
+
+  return dns.lookup(address, family, callback);
+}
+
+
+function lookup4(address, callback) {
+  return lookup(address || '0.0.0.0', 4, callback);
+}
+
+
+function newHandle(type) {
+  if (type == 'udp4') {
+    var handle = new UDP();
+    handle.lookup = lookup4;
+    return handle;
+  }
+
+  throw new Error('Bad socket type specified. Valid types are: udp4');
+}
+
+
+function Socket(type, listener) {
+  EventEmitter.call(this);
+
+  if (util.isObject(type)) {
+    var options = type;
+    type = options.type;
```

Comment:
Oh this is even more dangerous, since people will think this is local. Please move the "var options" declaration before the if to emphasize that is value is undefined by default.